documents/development/events.md

```diff
@@ -0,0 +1,108 @@
+- _occurredAt_ - Set by the client, optional. This likely differs from `createdAt` as it should store the local time of the event. Event’s may be batched when sent from the client, so we cannot rely on the time server received the event to be the time the event occurred. If there is no `occurredAt`, the server will apply `createdAt`
+- _type_ - Set by the client, required. The named type of the event. Examples include `pageview`, `item-added-to-cart`, `purchase`, or `click`
+- _userId_ - Set by the client, optional. If the client knows the primary identifying profile property of the current user, it should be sent. This is /your/ notion of userId. This property takes precedence over `anonymousId` when present.
+- _anonymousId_ - Set by the client, required (unless `userId` is present). If the `userId` is not known (the user is not yet logged in), a unique ID should be generated by the client. These can be device GUIDs in the case of mobile phones, or just another UUID.
+
+### `eventData` table
+
+This is a “has-many” data table logging the data properties of each event. Each event can have 0 or many rows in this table.
+
+- _guid_ - Set by the Grouparoo server. The primary Key for this portion of an event’s data. Generated by the Grouparoo server. Should /not/ be auto-incremented but others unique (we use UUIDv4) + a `edt_` prefix.
+- _key_- Set by the client, required. The named key of this portion of an event’s data.
+- _value_- Set by the client, required. The value of this portion of an event’s data. All data is stored as strings.
+
+### `apiKeys` table
+
+The `apiKeys` table is stored within Grouparoo’s database, regardless of the `EventBackend` in use. Event’s must be created with an an `apiKey` which is authorized to create events.
+
+## Workflow
+
+1. Create an `apiKey` (`_key` in our example) which is authorized to create events. We will give it the name “Web API”.
+2. A Client clicks on a Google add for our website and lands on our homepage.
+3. Since we know nothing about this user, we’ll create an `anonymousId=person_1` and send an event about the page view:
+
+```js
+Grouparoo.init({
+  apiKey: "_key",
+  endpoint: "https://your-grouparoo-instance.company.com", // likely pre-set by the client library
+});
+
+Gropuaroo.track({
+  anonymousId: "person_1",
+  occurredAt: new Date(), // likley a default in the Client Library
+  type: "pageview",
+  data: {
+    path: "/",
+    utm_source: "google",
+    utm_medium: "cpc",
+    utm_campaign: "my-ad-campaign",
+    // ...
+  },
+});
+```
+
+Grouparoo will create a new `Profile` to track all the activity we can for “person_1”, and attempt to decorate their profile with any additional information we can load from your `Profile Property Rules`.
+
+Eventually, after visiting a few more pages, the user will sign in, and we’ll know their `userId`
+
+```js
+Gropuaroo.track({
+  userId: 123,
+  occurredAt: new Date(),
```

Comment:
If the library doesn't handle silently sending the anonyous id with every request, they'd need to send it here to do the mapping. I think it's ok to always be sending it and we have the user `userId` take precedence.

In the [segment api](https://segment.com/docs/connections/spec/identify/) this is actually an `identify` call, but is to the same endpoint.

Reply:
But... if `userId` takes precedence... why require it?

Reply:
For a while, the user is `anonymousId: person-1` and then they login and now are `userId: 10`
If no event ever has both a `userId` and `anonymousId`, then Grouparoo doesn't know that `person-1` and `10` are the same profile. The `10` just looks like a completely separate logged in user to the server.